### PR TITLE
Support JWK as key format

### PR DESCRIPTION
**Pull request by woutslakhorst**
Using PEM formatted keys in json structures results in errors since newlines have to be added. Using JWK's solve this. It also helps in abstracting from the type of key (EC vs RSA).

Both the command line utility and the API's now support returning JWK's:

- the command line utility returns both JWK and PEM (was only PEM)
- the generateKeyPair and publicKey API's still return PEM by default
- the two API's will return JWK format when the Accept header in the http request equals `application/json`

### CODE REVIEW

**Comment by woutslakhorst** [comment]
The decrease in coverage is the result of very hard to hit error conditions

File: api/api.go [hunk 1]
```diff
@@ -38,13 +38,27 @@ type ApiWrapper struct {
...
 	le := types.LegalEntity{URI: string(params.LegalEntity)}
 	if err := w.C.GenerateKeyPairFor(le); err != nil {
 		return err
 	}
 
-	pub, err := w.C.PublicKey(le)
+	acceptHeader := ctx.Request().Header.Get("Accept")
+
+	// starts with so we can ignore any +
+	if strings.Index(acceptHeader, "application/json") == 0 {
```
Comment (flimzy) [inline api/api.go]: This can hit false positives (although admittedly, extremely rare ones). I.e. `application/json+some+proprietary+format` would validate as true.

My preference is to use [ParseMediaType](https://golang.org/pkg/mime/#ParseMediaType):

    if ct, _, _ := mime.ParseMediaType(acceptHeader); ct == "application/json" {

although it is a bit computationally heavy. If that's a concern (but is it, in a REST handler?), the stdlib in a few places splits on `;`, and considers only what occurs first.

File: api/api.go [hunk 2]
```diff
@@ -325,22 +339,41 @@ func (w *ApiWrapper) Verify(ctx echo.Context) error {
...
 	if match, err := regexp.MatchString(`^\S+$`, urn); !match || err != nil {
 		return echo.NewHTTPError(http.StatusBadRequest, "incorrect organization urn in request")
 	}
-	pubKey, err := w.C.PublicKey(types.LegalEntity{URI: urn})
 
+	le := types.LegalEntity{URI: urn}
+	acceptHeader := ctx.Request().Header.Get("Accept")
+
+	// starts with so we can ignore any +
+	if strings.Index(acceptHeader, "application/json") == 0 {
```
Comment (flimzy) [inline api/api.go]: Same comment here re: potential, rare false positives.

File: api/api.go [hunk 3]
```diff
@@ -325,22 +339,41 @@ func (w *ApiWrapper) Verify(ctx echo.Context) error {
...
+		jwk, err := w.C.PublicKeyInJWK(le)
+		if err != nil {
+			if strings.Contains(err.Error(), "could not open private key") {
```
Comment (flimzy) [inline api/api.go]: Can `PublicKeyInJWK` be made to return an expected error type, rather than depending on magic string heuristics?

This is usually a pretty simple change to implement.

File: api/api.go [hunk 4]
```diff
@@ -325,22 +339,41 @@ func (w *ApiWrapper) Verify(ctx echo.Context) error {
...
+				return ctx.NoContent(404)
+			}
+			logrus.Error(err.Error())
+			return err
```
Comment (flimzy) [inline api/api.go:362]: Why both logging and returning an error?

File: api/api.go [hunk 5]
```diff
@@ -325,22 +339,41 @@ func (w *ApiWrapper) Verify(ctx echo.Context) error {
...
+		}
+
+		return ctx.JSON(http.StatusOK, jwk)
+	}
+
+	// backwards compatible PEM format is the default
+	pub, err := w.C.PublicKeyInPEM(le)
 	if err != nil {
 		if strings.Contains(err.Error(), "could not open private key") {
```
Comment (flimzy) [inline api/api.go]: Not part of this PR, but same comment: Can `PublicKeyInPEM` be updated to return an expected error type, rather than magic string comparison?

File: api/api_test.go [hunk 6]
```diff
@@ -154,7 +185,7 @@ func TestApiWrapper_Encrypt(t *testing.T) {
 		echo.EXPECT().Request().Return(request)
 		echo.EXPECT().JSON(http.StatusOK, gomock.Any())
 
-		client.Encrypt(echo)
+		_ = client.Encrypt(echo)
```
Comment (flimzy) [inline api/api_test.go:188]: :+1:

File: api/api.go [hunk 7]
```diff
@@ -325,22 +339,41 @@ func (w *ApiWrapper) Verify(ctx echo.Context) error {
...
 			return ctx.NoContent(404)
 		}
-
 		logrus.Error(err.Error())
 		return err
 	}
 
-	return ctx.String(200, pubKey)
+	return ctx.String(http.StatusOK, pub)
```
Comment (flimzy) [inline api/api.go:378]: :+1:

File: engine/engine_test.go [hunk 8]
```diff
@@ -71,101 +72,78 @@ func TestNewCryptoEngine_Cmd(t *testing.T) {
...
-		cmd := e.Cmd
-		if cmd.Name() != "crypto" {
-			t.Errorf("Expected Cmd name to equal [crypto], got %s", cmd.Name())
-		}
-
-		if len(cmd.Commands()) != 3 {
-			t.Errorf("Expected Cmd to have 3 sub-command, got %d", len(cmd.Commands()))
-		}
+		assert.Equal(t, "crypto", cmd.Name())
+		assert.Len(t,  cmd.Commands(), 3)
```
Comment (flimzy) [inline engine/engine_test.go]: Personal note: I prefer the previous style over the assertion library, but I know you made a conscious choice to make the change, since you mentioned it in our meeting :)

File: engine/engine_test.go [hunk 9]
```diff
@@ -71,101 +72,78 @@ func TestNewCryptoEngine_Cmd(t *testing.T) {
...
 	})
 
 	t.Run("Running generateKeyPair with too few arguments gives error", func(t *testing.T) {
-		cmd := e.Cmd
-
 		cmd.SetArgs([]string{"generateKeyPair"})
-		cmd.SetOutput(new(bytes.Buffer))
+		cmd.SetOut(new(bytes.Buffer))
```
Comment (flimzy) [inline engine/engine_test.go]: I gather this is just a way to discard the output? I think `ioutil.Discard` would be a more idiomatic and performant approach.

**Review by flimzy — COMMENTED** [review]
I don't see any major issues here. Mostly minor idiomatic changes.

The only "big" thing I don't like here, is the use of the testify/assert library. I prefer to use the standard library for tests (I feel it hides less magic, and makes the tests more readable).  But that's very much a personal opinion--no objective truth to be found there. :)

File: api/api.go (hunk 1 above)
Comment (woutslakhorst) [reply api/api.go]: agreed

File: api/api.go (hunk 3 above)
Comment (woutslakhorst) [reply api/api.go]: has been rewritten, the check for the specific os error (eg: fileNotFound) is now being done at a lower level. Mainly to not check for an os specific error in the API layer.

File: api/api.go (hunk 4 above)
Comment (woutslakhorst) [reply api/api.go:362]: The error is returned to echo where it will do its magic to convert it into a 500 and some logging. The logging can be removed, but then we first have to configure the echo logger properly.

File: engine/engine_test.go (hunk 9 above)
Comment (woutslakhorst) [reply engine/engine_test.go]: replaced instances

File: engine/engine_test.go (hunk 8 above)
Comment (woutslakhorst) [reply engine/engine_test.go]: personally I find it more readable with the assertions, especially when doing:
```
if assert.NoError(t, err) {
 // extra assertions here
}
```

Plus the output of the assert library is more consistent than the different custom crafted error message from ourselves.

File: pkg/storage/fs.go [hunk 10]
```diff
@@ -32,6 +33,24 @@ import (
 
 const privateKeyFilePostfix = "private.pem"
 
+type FileOpenError struct {
+	filePath 	string
+	legalEntity string
```
Comment (flimzy) [inline pkg/storage/fs.go:38]: Inconsistent indentation... gofmt snafu?

File: engine/engine_test.go (hunk 8 above)
Comment (flimzy) [reply engine/engine_test.go]: I can understand that. What I most dislike about an assertion library is that some of the assertions are non-obvious.  There's never room for confusion with a simple `if` statement.

And for an error case like that, I usually handle it with something like:

```
    err := somethingThatShouldNotFail()
    if err != nil {
        t.Fatal(err)
    }
    // extra assertions here
```

I believe that accomplishes the same thing. It is two more lines of code, but I don't mind that for a little extra readability.

But again, this is only my opinion. I know you've made the choice to go with an assertion library, and I'm not trying to dissuade you from that. Just offering my honest review :)

File: pkg/storage/fs.go (hunk 10 above)
Comment (woutslakhorst) [reply pkg/storage/fs.go:38]: This triggered me to update some of my editor settings to do auto gofmt and goimports
